README.md
# gregorian [![docs][docs-badge]][docs] [![tests][tests-badge]][tests]
[docs]: https://docs.rs/gregorian/
[tests]: https://github.com/de-vri-es/gregorian-rs/actions?query=workflow%3Atests
[docs-badge]: https://docs.rs/gregorian/badge.svg
[tests-badge]: https://github.com/de-vri-es/gregorian-rs/workflows/tests/badge.svg

An implementation of the proleptic Gregorian calendar, compatible with ISO 8601.
Amongst others, that means that the calendar has a year zero preceeding the year 1.

This create does not deal with times or time zones.

The [`Date`] type represents a date (year, month and day),
the [`Year`] type represents a calendar year,
the [`Month`] type represents a calendar month,
and the [`YearMonth`] type represents a month of a specific year.

Where possible, things are implemented as `const fn`.
Currently, this excludes trait implementations and functions that rely on traits.

## Example
```rust
use gregorian::{Date, Month::*, Year, YearMonth};

assert!(Year::new(2020).has_leap_day(), true);
assert!(YearMonth::new(1900, February).total_days() == 28);
assert!(YearMonth::new(2000, February).total_days() == 29);

assert!(Year::new(2020).with_month(March).first_day() == Date::new(2020, March, 1).unwrap());
assert!(Year::new(2020).with_month(March).last_day() == Date::new(2020, March, 31).unwrap());

assert!(Year::new(2020).first_day() == Date::new(2020, 1, 1).unwrap());
assert!(Year::new(2020).last_day() == Date::new(2020, 12, 31).unwrap());

assert!(Date::new(2020, 2, 1).unwrap().day_of_year() == 32);
```

## Rounding invalid dates
When you use [`Date::add_years()`] or [`Date::add_months()`], you can get invalid dates.
These are reported with an [`InvalidDayOfMonth`] error which has the
[`next_valid()`][InvalidDayOfMonth::next_valid] and [`prev_valid()`][InvalidDayOfMonth::prev_valid] methods.
Those can be used to get the next or previous valid date instead.

Additionally, there is an extension trait for `Result<Date, InvalidDayOfMonth>` with the
[`or_next_valid()`][DateResultExt::or_next_valid] and [`or_prev_valid()`][DateResultExt::or_prev_valid] methods.
This allows you to directly round the date on the `Result` object.

```rust
use gregorian::{Date, DateResultExt};
let date = Date::new(2020, 1, 31).unwrap();
assert!(date.add_months(1).or_next_valid() == Date::new(2020, 3, 1).unwrap());
assert!(date.add_months(1).or_prev_valid() == Date::new(2020, 2, 29).unwrap());
```

[`Date`]: https://docs.rs/gregorian/latest/gregorian/struct.Date.html
[`Year`]: https://docs.rs/gregorian/latest/gregorian/struct.Year.html
[`YearMonth`]: https://docs.rs/gregorian/latest/gregorian/struct.YearMonth.html
[`Month`]: https://docs.rs/gregorian/latest/gregorian/struct.Month.html
[`InvalidDayOfMonth`]: https://docs.rs/gregorian/latest/gregorian/struct.InvalidDayOfMonth.html

[`Date::add_years()`]: https://docs.rs/gregorian/latest/gregorian/struct.Date.html#method.add_years
[`Date::add_months()`]: https://docs.rs/gregorian/latest/gregorian/struct.Date.html#method.add_months
[DateResultExt::or_next_valid]: https://docs.rs/gregorian/latest/gregorian/trait.DateResultExt.html#method.or_next_valid
[DateResultExt::or_prev_valid]: https://docs.rs/gregorian/latest/gregorian/trait.DateResultExt.html#method.or_prev_valid
[InvalidDayOfMonth::next_valid]: https://docs.rs/gregorian/latest/gregorian/struct.InvalidDayOfMonth.html#method.next_valid
[InvalidDayOfMonth::prev_valid]: https://docs.rs/gregorian/latest/gregorian/struct.InvalidDayOfMonth.html#method.prev_valid


CHANGELOG.md
Version 0.2.4 - 2022-11-14:
  * Add optional `serde` support.

Version 0.2.3 - 2022-05-24:
  * Tweak signature of `Date::days_since()`.

Version 0.2.2 - 2022-05-24 (YANKED):
  * Fix `Date::today()` for Windows.
  * Add `date.days_since(other)` to compute difference between dates in days.

Version 0.2.1 - 2021-01-24:
  * Add `Date::today()` and `Date::today_utc()`.

Version 0.2.0 - 2020-10-17:
  * Make `Debug` string representations more compact.
  * `Month::wrapping_add/sub` now take an `i8` instead of `u8`.
  * Add `YearMonth::add/sub_years` and `YearMonth::add/sub_months`.
  * Add `Date::add/sub_years` and `Date::add/sub_months`.
  * Add `InvalidDayOfMonth::next/prev_valid()` to get the nearest valid date.
  * Add `DateResultExt` extension trait for `Result<Date, InvalidDayOfMonth>` to easily round invalid dates.
  * Rename `InvalidDate::InvalidDayForMonth` to `InvalidDate::InvalidDayOfMonth` to match the wrapped struct.
  * Rename `InvalidDayOfYear::day` to `day_of year`.
  * Make many functions `const`.

Version 0.1.1 - 2020-10-16:
  * Fix `Date::add_days` for days in years 100, 200 and 300 of the 400 year cycle.

Version 0.1.0 - 2020-06-20:
  * Initial release.


src/year_month.rs
use crate::{Date, InvalidDayOfMonth, Month, Year};

/// A month of a specific year.
#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct YearMonth {
	year: Year,
	month: Month,
}

impl YearMonth {
	/// Create a new year-month.
	pub fn new(year: impl Into<Year>, month: Month) -> Self {
		let year = year.into();
		Self { year, month }
	}

	/// Create a new year-month at compile time.
	///
	/// Unline, [`Self::new()`], this takes a [`Year`] instead of an `impl Into<Year>`,
	/// since the conversion can not be done in a `const fn`.
	pub const fn new_const(year: Year, month: Month) -> Self {
		Self { year, month }
	}

	/// Get the year.
	pub const fn year(self) -> Year {
		self.year
	}

	/// Get the month as [`Month`].
	pub const fn month(self) -> Month {
		self.month
	}

	/// Get the total number of days in the month.
	///
	/// This function accounts for leap-days,
	/// so it reports 29 days for February of leap-years,
	/// and 28 days for other years.
	pub const fn total_days(self) -> u8 {
		crate::raw::days_in_month(self.month, self.year.has_leap_day())
	}

	/// Get the day-of-year on which the month starts.
	///
	/// Day-of-year numbers are 1-based.
	pub const fn day_of_year(self) -> u16 {
		crate::raw::start_day_of_year(self.month, self.year.has_leap_day())
	}

	/// Get the next month as [`YearMonth`].
	///
	/// After December, this function returns January of the next year.
	pub const fn next(self) -> Self {
		if let Month::December = self.month {
			Self::new_const(self.year.next(), Month::January)
		} else {
			Self::new_const(self.year, self.month.wrapping_next())
		}
	}

	/// Get the previous month as [`YearMonth`].
	///
	/// After January, this function returns December of the previous year.
	pub const fn prev(self) -> Self {
		if let Month::January = self.month {
			Self::new_const(self.year.prev(), Month::December)
		} else {
			Self::new_const(self.year, self.month.wrapping_prev())
		}
	}

	/// Get a new [`YearMonth`] by adding a number of years.
	pub const fn add_years(self, years: i16) -> Self {
		let year = Year::new(self.year.to_number() + years);
		year.with_month(self.month())
	}

	/// Get a new [`YearMonth`] by subtracting a number of years.
	pub const fn sub_years(self, years: i16) -> Self {
		let year = Year::new(self.year.to_number() - years);
		year.with_month(self.month())
	}

	/// Get a new [`YearMonth`] by adding a number of months.
	pub const fn add_months(self, months: i32) -> Self {
		// Split calculation for years and months.
		let months = (self.month().to_number() - 1) as i32 + months;
		let mut year = self.year().to_number() + (months / 12) as i16;
		let month = Month::January.wrapping_add((months % 12) as i8);

		// If we subtract months, we must decrease the year too.
		if months % 12 < 0 {
			year -= 1;
		}

		Year::new(year).with_month(month)
	}

	/// Get a new [`YearMonth`] by subtracting a number of months.
	pub const fn sub_months(self, months: i32) -> Self {
		// This breaks for i32::MIN, but that would overflow the year counter anyway.
		self.add_months(-months)
	}

	/// Combine the year and month with a day, to create a full [`Date`].
	pub const fn with_day(self, day: u8) -> Result<Date, InvalidDayOfMonth> {
		if let Err(e) = InvalidDayOfMonth::check(self.year, self.month, day) {
			return Err(e);
		}
		unsafe { Ok(Date::new_unchecked(self.year, self.month, day)) }
	}

	/// Combine the year and month with a day, without checking for validity.
	///
	/// # Safety
	/// Although this is currently not the case, future implementations may rely on date validity for memory safety
	pub const unsafe fn with_day_unchecked(self, day: u8) -> Date {
		Date::new_unchecked(self.year, self.month, day)
	}

	/// Get the first day of the month as [`Date`].
	pub const fn first_day(self) -> Date {
		Date {
			year: self.year,
			month: self.month,
			day: 1,
		}
	}

	/// Get the last day of the month as [`Date`].
	pub const fn last_day(self) -> Date {
		Date {
			year: self.year,
			month: self.month,
			day: self.total_days(),
		}
	}
}

impl core::fmt::Display for YearMonth {
	fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
		write!(f, "{:04}-{:02}", self.year.to_number(), self.month().to_number())
	}
}

impl core::fmt::Debug for YearMonth {
	fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
		write!(f, "YearMonth({})", self)
	}
}

#[cfg(test)]
mod test {
	use crate::*;
	use assert2::{assert, let_assert};

	#[test]
	fn add_months() {
		for i in -200..=200 {
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 1) == Year::new(2000 + i as i16).with_month(February));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 2) == Year::new(2000 + i as i16).with_month(March));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 3) == Year::new(2000 + i as i16).with_month(April));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 4) == Year::new(2000 + i as i16).with_month(May));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 5) == Year::new(2000 + i as i16).with_month(June));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 6) == Year::new(2000 + i as i16).with_month(July));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 7) == Year::new(2000 + i as i16).with_month(August));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 8) == Year::new(2000 + i as i16).with_month(September));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 9) == Year::new(2000 + i as i16).with_month(October));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 10) == Year::new(2000 + i as i16).with_month(November));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 11) == Year::new(2000 + i as i16).with_month(December));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + 12) == Year::new(2001 + i as i16).with_month(January));

			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -1) == Year::new(1999 + i as i16).with_month(December));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -2) == Year::new(1999 + i as i16).with_month(November));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -3) == Year::new(1999 + i as i16).with_month(October));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -4) == Year::new(1999 + i as i16).with_month(September));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -5) == Year::new(1999 + i as i16).with_month(August));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -6) == Year::new(1999 + i as i16).with_month(July));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -7) == Year::new(1999 + i as i16).with_month(June));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -8) == Year::new(1999 + i as i16).with_month(May));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -9) == Year::new(1999 + i as i16).with_month(April));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -10) == Year::new(1999 + i as i16).with_month(March));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -11) == Year::new(1999 + i as i16).with_month(February));
			assert!(Year::new(2000).with_month(January).add_months(i * 12 + -12) == Year::new(1999 + i as i16).with_month(January));
		}
	}

	#[test]
	fn total_days() {
		assert!(Year::new(2020).with_month(January).total_days() == 31);
		assert!(Year::new(2020).with_month(February).total_days() == 29);
		assert!(Year::new(2020).with_month(March).total_days() == 31);
		assert!(Year::new(2020).with_month(April).total_days() == 30);
		assert!(Year::new(2020).with_month(May).total_days() == 31);
		assert!(Year::new(2020).with_month(June).total_days() == 30);
		assert!(Year::new(2020).with_month(July).total_days() == 31);
		assert!(Year::new(2020).with_month(August).total_days() == 31);
		assert!(Year::new(2020).with_month(September).total_days() == 30);
		assert!(Year::new(2020).with_month(October).total_days() == 31);
		assert!(Year::new(2020).with_month(November).total_days() == 30);
		assert!(Year::new(2020).with_month(December).total_days() == 31);

		assert!(Year::new(2021).with_month(January).total_days() == 31);
		assert!(Year::new(2021).with_month(February).total_days() == 28);
		assert!(Year::new(2021).with_month(March).total_days() == 31);
		assert!(Year::new(2021).with_month(April).total_days() == 30);
		assert!(Year::new(2021).with_month(May).total_days() == 31);
		assert!(Year::new(2021).with_month(June).total_days() == 30);
		assert!(Year::new(2021).with_month(July).total_days() == 31);
		assert!(Year::new(2021).with_month(August).total_days() == 31);
		assert!(Year::new(2021).with_month(September).total_days() == 30);
		assert!(Year::new(2021).with_month(October).total_days() == 31);
		assert!(Year::new(2021).with_month(November).total_days() == 30);
		assert!(Year::new(2021).with_month(December).total_days() == 31);
	}

	#[test]
	fn start_day_of_year() {
		for year in -400..=400 {
			// We assume total_days() works, since it has it's own unit test,
			// then cross check start days against that.
			let mut start_day = 1;
			for month in &Year::new(year).months() {
				assert!(month.day_of_year() == start_day);
				start_day += u16::from(month.total_days());
			}
		}
	}

	#[test]
	#[cfg(feature = "std")]
	fn format() {
		assert!(format!("{}", Year::new(2020).with_month(January)) == "2020-01");
		assert!(format!("{:?}", Year::new(2020).with_month(January)) == "YearMonth(2020-01)");
	}

	#[test]
	fn serde() {
		let_assert!(Ok(serialized) = serde_yaml::to_string(&YearMonth::new(2020, Month::January)));
		assert!(serialized == "year: 2020\nmonth: 1\n");

		let_assert!(Ok(parsed) = serde_yaml::from_str::<YearMonth>("year: 2020\nmonth: 1\n"));
		assert!(parsed.year == 2020);
		assert!(parsed.month == Month::January);
	}
}
